Druid/DRCogitStackToRegisterFlushHoisting.class.st

```diff
@@ -0,0 +1,31 @@
+Class {
+	#name : #DRCogitStackToRegisterFlushHoisting,
+	#superclass : #DROptimisation,
+	#category : #'Druid-BytecodeToJITCompilation'
+}
+
+{ #category : #accessing }
+DRCogitStackToRegisterFlushHoisting >> applyTo: cfg [
+
+	"Insert a flush stack instruction in the common dominator of all those instructions that require stack flushing.
+	Explain why we need to hoist it."
+	
+	| toMerge dominator marshall instructionsNeedingStackFlush |
+	instructionsNeedingStackFlush := OrderedCollection new.
+	cfg instructionsDo: [ :e |
+		e needsStackFlush ifTrue: [ instructionsNeedingStackFlush add: e ]
+	].
+	
+	"If no instructions require flushing, do nothing"
+	instructionsNeedingStackFlush ifEmpty: [ ^ self ].
+	
+	"If only one instruction requires, flush just before the instruction"
+	instructionsNeedingStackFlush size = 1 ifTrue: [
+		instructionsNeedingStackFlush first addBefore: cfg instructionFactory flushStack.
+		^ self
+	].
```

Comment:
This special case would not be necessary if the flush is added always at the top of the dominator block.

Reply:
But I'm not sure if It's correct to do that because we may have stack manipulation instructions in the middle.
The only way to make it correct is maybe to split the basic blocks on these instructions requiring stack flushing. Then putting it at the top of the block will always be putting it just before the instruction we care about :)